br/pkg/lightning/restore/check_info.go

```diff
@@ -399,535 +180,64 @@ func (rc *Controller) estimateSourceData(ctx context.Context) (int64, int64, boo
-		for i := len(row); i < colCountFromTiDB; i++ {
-			if _, ok := defaultCols[core.Columns[i].Name.L]; !ok {
-				msgs = append(msgs, fmt.Sprintf("TiDB schema `%s`.`%s` has %d columns,"+
-					"and data file has %d columns, but column %s are missing the default value,"+
-					"please give column a default value to skip this check",
-					tableInfo.DB, tableInfo.Name, colCountFromTiDB, len(row), core.Columns[i].Name.L))
-			}
-		}
-		return msgs, nil
-	}
-
-	// compare column names and make sure
-	// 1. TiDB table info has data file's all columns(besides ignore columns)
-	// 2. Those columns not introduced in data file always have a default value.
-	colMap := make(map[string]struct{})
-	for col := range igCols {
-		colMap[col] = struct{}{}
-	}
-	for _, col := range core.Columns {
-		if _, ok := colMap[col.Name.L]; ok {
-			// tidb's column is ignored
-			// we need ensure this column has the default value.
-			if _, hasDefault := defaultCols[col.Name.L]; !hasDefault {
-				msgs = append(msgs, fmt.Sprintf("TiDB schema `%s`.`%s`'s column %s cannot be ignored,"+
-					"because it doesn't have a default value, please set tables.ignoreColumns properly",
-					tableInfo.DB, tableInfo.Name, col.Name.L))
-			}
-		} else {
-			colMap[col.Name.L] = struct{}{}
-		}
-	}
-	// tidb_rowid can be ignored in check
-	colMap[model.ExtraHandleName.String()] = struct{}{}
-	for _, col := range colsFromDataFile {
-		if _, ok := colMap[col]; !ok {
-			checkMsg := "please check table schema"
-			if dataFileMeta.Type == mydump.SourceTypeCSV && rc.cfg.Mydumper.CSV.Header {
-				checkMsg += " and csv file header"
-			}
-			msgs = append(msgs, fmt.Sprintf("TiDB schema `%s`.`%s` doesn't have column %s, "+
-				"%s or use tables.ignoreColumns to ignore %s",
-				tableInfo.DB, tableInfo.Name, col, checkMsg, col))
-		} else {
-			// remove column for next iteration
-			delete(colMap, col)
-		}
+		return errors.Trace(err)
 	}
-	// if theses rest columns don't have a default value.
-	for col := range colMap {
-		if _, ok := defaultCols[col]; ok {
-			continue
-		}
-		msgs = append(msgs, fmt.Sprintf("TiDB schema `%s`.`%s` doesn't have the default value for %s. "+
-			"Please add default value for column '%s' or choose another column to ignore or add this column in data file",
-			tableInfo.DB, tableInfo.Name, col, col))
+	if result != nil {
+		rc.checkTemplate.Collect(result.Severity, result.Passed, result.Message)
 	}
-	return msgs, nil
+	return nil
 }
 
-// checkCSVHeader try to check whether the csv header config is consistent with the source csv files by:
-// 1. pick one table with two CSV files and a unique/primary key
-// 2. read the first row of those two CSV files
-// 3. checks if the content of those first rows are compatible with the table schema, and whether the
-//    two rows are identical, to determine if the first rows are a header rows.
-func (rc *Controller) checkCSVHeader(ctx context.Context, dbMetas []*mydump.MDDatabaseMeta) error {
```

Comment:
Extracted as `(*csvHeaderCheckItem).Check()`, in `br/pkg/lightning/restore/precheck_impl.go`, L860